File: evaluation/S0/javascript/Hard/1905-shortest-subarray-with-sum-at-least-k.js
// Source: https://leetcode.com/problems/shortest-subarray-with-sum-at-least-k/   |   Difficulty: Hard
//
// Problem Description:
// Given an integer array nums and an integer k, return the length of the shortest non-empty subarray of nums with a sum of at least k. If there is no such subarray, return -1.
//
// A subarray is a contiguous part of an array.
//
// Example:
// Input: nums = [1], k = 1
// Output: 1
//
// Constraints:
// 1 <= nums.length <= 105
// 	-105 <= nums[i] <= 105
// 	1 <= k <= 109
//

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var shortestSubarray = function(nums, k) {
    const n = nums.length;
    const prefixSums = new Array(n + 1).fill(0);
    for (let i = 0; i < n; i++) {
        prefixSums[i + 1] = prefixSums[i] + nums[i];
    }
    const deque = [];
    let minLength = Infinity;

    for (let i = 0; i <= n; i++) {
        while (deque.length > 0 && prefixSums[i] - prefixSums[deque[0]] >= k) {
            minLength = Math.min(minLength, i - deque.shift());
        }
        while (deque.length > 0 && prefixSums[i] <= prefixSums[deque[deque.length - 1]]) {
            deque.pop();
        }
        deque.push(i);
    }
    return minLength === Infinity ? -1 : minLength; 
};

File: evaluation/S0/python/Hard/1905-shortest-subarray-with-sum-at-least-k.py
# Source: https://leetcode.com/problems/shortest-subarray-with-sum-at-least-k/   |   Difficulty: Hard
#
# Problem Description:
# Given an integer array nums and an integer k, return the length of the shortest non-empty subarray of nums with a sum of at least k. If there is no such subarray, return -1.
#
# A subarray is a contiguous part of an array.
#
# Example:
# Input: nums = [1], k = 1
# Output: 1
#
# Constraints:
# 1 <= nums.length <= 105
# 	-105 <= nums[i] <= 105
# 	1 <= k <= 109
#

class Solution(object):
    def shortestSubarray(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        from collections import deque
        n = len(nums)
        prefix_sums = [0] * (n + 1)
        
        for i in range(n):
            prefix_sums[i + 1] = prefix_sums[i] + nums[i]
        
        dq = deque()
        min_length = float('inf')
        
        for i in range(n + 1):
            while dq and prefix_sums[i] - prefix_sums[dq[0]] >= k:
                min_length = min(min_length, i - dq.popleft())
            
            while dq and prefix_sums[i] <= prefix_sums[dq[-1]]:
                dq.pop()
            
            dq.append(i)
        
        return min_length if min_length != float('inf') else -1